File: app/src/main/java/cn/surine/schedulex/ui/schedule_data_export/ScheduleDataExport.java
package cn.surine.schedulex.ui.schedule_data_export;

import android.Manifest;
import android.annotation.SuppressLint;

import androidx.lifecycle.ViewModelProviders;

import com.google.android.material.snackbar.Snackbar;
import com.tbruyelle.rxpermissions2.RxPermissions;

import java.util.ArrayList;
import java.util.List;

import cn.surine.schedulex.R;
import cn.surine.schedulex.base.controller.BaseBindingFragment;
import cn.surine.schedulex.base.utils.CalendarProviders;
import cn.surine.schedulex.base.utils.Dates;
import cn.surine.schedulex.base.utils.Files;
import cn.surine.schedulex.base.utils.InstanceFactory;
import cn.surine.schedulex.base.utils.Jsons;
import cn.surine.schedulex.base.utils.Objs;
import cn.surine.schedulex.base.utils.Toasts;
import cn.surine.schedulex.data.entity.Course;
import cn.surine.schedulex.data.entity.Schedule;
import cn.surine.schedulex.databinding.FragmentDateExportBinding;
import cn.surine.schedulex.ui.course.CourseRepository;
import cn.surine.schedulex.ui.course.CourseViewModel;
import cn.surine.schedulex.ui.schedule.ScheduleRepository;
import cn.surine.schedulex.ui.schedule.ScheduleViewModel;
import cn.surine.schedulex.ui.schedule_config.ScheduleConfigFragment;

/**
 * Intro：
 *
 * @author sunliwei
 * @date 2020-02-05 15:14
 */
public class ScheduleDataExport extends BaseBindingFragment<FragmentDateExportBinding> {
    private Schedule schedule;
    private CourseViewModel courseViewModel;

    @Override
    public int layoutId() {
        return R.layout.fragment_date_export;
    }

    @SuppressLint("CheckResult")
    @Override
    protected void onInit(FragmentDateExportBinding t) {
        if(Objs.isNull(getArguments())){
            return;
        }

        Class[] classes = new Class[]{ScheduleRepository.class};
        Object[] args = new Object[]{ScheduleRepository.abt.getInstance()};
        ScheduleViewModel scheduleViewModel = ViewModelProviders.of(this, InstanceFactory.getInstance(classes, args)).get(ScheduleViewModel.class);

        Class[] classesForCourse = new Class[]{CourseRepository.class};
        Object[] argsForCourse = new Object[]{CourseRepository.abt.getInstance()};
        courseViewModel = ViewModelProviders.of(this, InstanceFactory.getInstance(classesForCourse, argsForCourse)).get(CourseViewModel.class);


        int scheduleId = getArguments().getInt(ScheduleConfigFragment.SCHEDULE_ID);
        schedule = scheduleViewModel.getScheduleById(scheduleId);
        t.scheduleTitle.setText(schedule.name);

        t.exportIcs.setOnClickListener(v -> {
            RxPermissions permission = new RxPermissions(activity());
            permission.request(Manifest.permission.READ_CALENDAR,Manifest.permission.WRITE_CALENDAR).subscribe(aBoolean -> {
                if (aBoolean) {
                    exportIcs();
                } else {
                    Toasts.toast(getString(R.string.permission_is_denied));
                }
            });
        });


        //导出为Json
        t.exportJson.setOnClickListener(v -> {
            RxPermissions rxPermissions = new RxPermissions(activity());
            rxPermissions.request(Manifest.permission.WRITE_EXTERNAL_STORAGE,Manifest.permission.READ_EXTERNAL_STORAGE).subscribe(aBoolean -> {
                if (aBoolean) {
                    saveToJson(t,scheduleId);
                } else {
                    Toasts.toast(getString(R.string.permission_is_denied));
                }
            });
        });

        t.exportExcel.setOnClickListener(v -> Toasts.toast("暂不支持！"));
        t.other.setOnClickListener(v -> Toasts.toast("欢迎加群提出意见！"));
    }


    /**
     * 导出到日历
     * */
    private void exportIcs() {
        List<Course> data = courseViewModel.getCourseByScheduleId(schedule.roomId);
        for (int i = 0; i < data.size(); i++) {
            Course course = data.get(i);
            List<Integer> weeks = new ArrayList<>();
            for (int j = 0; j < course.classWeek.length(); j++) {
                if(course.classWeek.charAt(i) == '1'){
                    weeks.add(j + 1);
                }
            }
            for (int j = 0; j < weeks.size(); j++) {
                //只记录比当前周大的数据
                if(weeks.get(j) >= schedule.curWeek()){
                    long startTime = getCalendarTime(weeks.get(j),course.classDay);
                    long endTime = startTime + 45 * 60 * 1000;
//                    CalendarProviders.addEvent(activity(),course.coureName,course.teachingBuildingName + course.classroomName,startTime,endTime);
                    CalendarProviders.deleteCalendarEvent(activity(),course.coureName);
                }
            }
        }
    }


    /**
     * 求得起始时间
     * */
    private long getCalendarTime(int week, String classDay) {
        return Dates.getDate(schedule.termStartDate,Dates.yyyyMMdd).getTime() + (7 * (week - 1) + Integer.parseInt(classDay) - 1) * Dates.ONE_DAY;
    }

    /**
     * 保存成json
     * @param scheduleId
     * @param t
     * */
    private void saveToJson(FragmentDateExportBinding t, int scheduleId) {
        String fileName = schedule.name + "_" + scheduleId;
        if(Files.saveAsJson(fileName, Jsons.entityToJson(courseViewModel.getCourseByScheduleId(scheduleId)))){
            Snackbar.make(t.exportJson,"保存成功,路径 /Download/"+fileName+".json",Snackbar.LENGTH_SHORT).show();
        }else{
            Toasts.toast("保存失败！请稍后再试！");
        }
    }
}

File: app/src/main/java/cn/surine/schedulex/ui/schedule_data_export/ScheduleDataExport.kt
package cn.surine.schedulex.ui.schedule_data_export

import android.Manifest
import android.app.ProgressDialog
import android.view.View
import cn.surine.schedulex.R
import cn.surine.schedulex.app_base.VmManager
import cn.surine.schedulex.app_base.hit
import cn.surine.schedulex.base.controller.BaseFragment
import cn.surine.schedulex.base.utils.*
import cn.surine.schedulex.data.entity.Course
import cn.surine.schedulex.data.entity.Schedule
import cn.surine.schedulex.ui.course.CourseViewModel
import cn.surine.schedulex.ui.schedule.ScheduleViewModel
import cn.surine.schedulex.ui.schedule_import_pro.model.CourseWrapper
import cn.surine.schedulex.ui.schedule_list.ScheduleListFragment.Companion.SCHEDULE_ID
import cn.surine.schedulex.ui.timetable_list.TimeTableViewModel
import com.google.android.material.snackbar.Snackbar
import com.tbruyelle.rxpermissions2.RxPermissions
import kotlinx.android.synthetic.main.fragment_date_export.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.*
import kotlin.collections.ArrayList

/**
 * Intro：
 *
 * @author sunliwei
 * @date 2020/6/26 10:47
 */
class ScheduleDataExport : BaseFragment() {

    lateinit var scheduleViewModel: ScheduleViewModel
    lateinit var courseViewModel: CourseViewModel
    private lateinit var timeTableViewModel: TimeTableViewModel
    lateinit var schedule: Schedule
    val calendarIds = ArrayList<Long>()
    override fun layoutId(): Int = R.layout.fragment_date_export

    override fun onInit(parent: View?) {
        arguments ?: return
        VmManager(this).apply {
            scheduleViewModel = vmSchedule
            courseViewModel = vmCourse
            timeTableViewModel = vmTimetable
        }
        val scheduleId = requireArguments().getInt(SCHEDULE_ID)
        schedule = scheduleViewModel.getScheduleById(scheduleId.toLong())
        scheduleTitle.text = schedule.name

        exportJson.setOnClickListener {
            RxPermissions(activity()).apply {
                request(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE).subscribe {
                    if (it) {
                        hit("save_json")
                        saveToJson(scheduleId)
                    } else {
                        Toasts.toast(getString(R.string.permission_is_denied))
                    }
                }
            }
        }

        other.setOnLongClickListener {
            RxPermissions(activity()).apply {
                request(Manifest.permission.READ_CALENDAR, Manifest.permission.WRITE_CALENDAR).subscribe {
                    if (it) {
                        hit("delete_calendar")
                        CoroutineScope(Dispatchers.IO).launch {
                            deleteCourseTask()
                        }
                    } else {
                        Toasts.toast(getString(R.string.permission_is_denied))
                    }
                }
            }
            true
        }
        other.setOnClickListener {
            RxPermissions(activity()).apply {
                request(Manifest.permission.READ_CALENDAR, Manifest.permission.WRITE_CALENDAR).subscribe {
                    if (it) {
                        hit("add_calendar")
                        val dialog = ProgressDialog(activity()).apply {
                            setMessage("正在导出至日历，请勿退出~ (如需删除，请长按导出按钮)")
                            setTitle("提醒")
                            setCancelable(false)
                            show()
                        }
                        CoroutineScope(Dispatchers.IO).launch {
                            addCourseTask(scheduleId, dialog)
                        }
                    } else {
                        Toasts.toast(getString(R.string.permission_is_denied))
                    }
                }
            }
        }

        exportExcel.setOnClickListener { Toasts.toast("暂不支持！") }
    }

    /**
     * 删除操作
     * */
    private suspend fun deleteCourseTask() {
        withContext(Dispatchers.Default) {
            Calendars.removeAllEvent(activity(), activity().packageName)
        }
        withContext(Dispatchers.Main) {
            Toasts.toast("删除成功！")
        }
    }

    /**
     * 添加日程
     * */
    private suspend fun addCourseTask(scheduleId: Int, dialog: ProgressDialog) {
        withContext(Dispatchers.Default) {
            val list: List<Course> = courseViewModel.getCourseByScheduleId(scheduleId)
            val c: Calendar = Calendar.getInstance()
            c.timeInMillis = System.currentTimeMillis()
            for (course in list) {
                //事件名称和提示信息
                val eventTitle = course.coureName
                val addrs = course.teachingBuildingName + course.classroomName
                val eventMsg = addrs + "/" + course.teacherName
                //当前周
                val curSchedule = scheduleViewModel.curSchedule
                val currentWeek = curSchedule.curWeek()
                val termStartTime = Dates.getDate(curSchedule.termStartDate, Dates.yyyyMMdd).time
                for ((index, week) in course.classWeek.withIndex()) {
                    //历史的周不会再添加
                    if (index + 1 < currentWeek || week == '0')
                        continue
                    val dateOffSet = termStartTime + Dates.ONE_DAY * (index * 7 + course.classDay.toInt() - 1)
                    val eventStartTime = dateOffSet + getCourseTime(course.classSessions.toInt(), true)
                    val eventEndTime = dateOffSet + getCourseTime(course.classSessions.toInt() + course.continuingSession.toInt() - 1, false)
                    //insert into calendar
                    val result = Calendars.addCalendarEvent(activity(), eventTitle, eventMsg, eventStartTime, eventEndTime, addr = addrs)
                    if (result != -1L) {
                        calendarIds.add(result)
                    }
                }
            }
        }
        withContext(Dispatchers.Main) {
            dialog.dismiss()
            Toasts.toast("添加成功！")
        }
    }


    /**
     * 获取上课时间
     * */
    private fun getCourseTime(classSessions: Int, start: Boolean): Long {
        val timeTable = timeTableViewModel.getTimTableById(scheduleViewModel.curSchedule.timeTableId)
        timeTable ?: return 0
        val bTimeTable = DataMaps.dataMappingTimeTableToBTimeTable(timeTable)
        if (classSessions > bTimeTable.timeInfoList.size) return 0
        val time = if (start) bTimeTable.timeInfoList[classSessions - 1].startTime else bTimeTable.timeInfoList[classSessions - 1].endTime
        return Dates.getTransformTimeString(time) * 60 * 1000
    }


    private fun saveToJson(scheduleId: Int) {
        val fileName = "${schedule.name}_$scheduleId"
        val courses = courseViewModel.getCourseByScheduleId(scheduleId)
        val targetList = mutableListOf<CourseWrapper>()
        for (i in courses) {
            targetList.add(DataMaps.dataMappingCourse2CourseWrapper(i))
        }
        if (Files.saveAsJson(fileName, Jsons.entityToJson(targetList))) {
            Snackbar.make(exportJson, "保存成功,路径 /Download/$fileName.json", Snackbar.LENGTH_SHORT).show();
        } else {
            Toasts.toast("保存失败！请稍后再试！");
        }
    }


}
